server/app/uk/gov/ons/addressIndex/server/controllers/AddressController.scala

```diff
@@ -128,68 +130,85 @@ class AddressController @Inject()(
+            inputAddress = address.inputAddress,
+            matchedFormattedAddress = address.matchedFormattedAddress,
+            id = address.id,
+            organisationName = address.tokens.getOrElse(Tokens.organisationName, ""),
+            departmentName = address.tokens.getOrElse(Tokens.departmentName, ""),
+            subBuildingName = address.tokens.getOrElse(Tokens.subBuildingName, ""),
+            buildingName = address.tokens.getOrElse(Tokens.buildingName, ""),
+            buildingNumber = address.tokens.getOrElse(Tokens.buildingNumber, ""),
+            streetName = address.tokens.getOrElse(Tokens.streetName, ""),
+            locality = address.tokens.getOrElse(Tokens.locality, ""),
+            townName = address.tokens.getOrElse(Tokens.townName, ""),
+            postcode = address.tokens.getOrElse(Tokens.postcode, ""),
+            uprn = address.hybridAddress.uprn,
+            score = address.hybridAddress.score
+          )
+        }
       )
+    )
   }
 
+  /**
+    * Iterates over requests data and adapts the size of the bulk-chunks using back-pressure.
+    * ES rejects requests when it cannot handle them (because of the consumed resources)
+    * in this case we reduce the bulk size so that we could process the data successfully.
+    * Otherwise we increase the size of the bulk so that we could do more in one bulk
+    *
+    * It should throw an exception if the situation is desperate (we only do one request at
+    * a time and this request fails)
+    * @param requests Stream of data that will be used to query ES
+    * @param miniBatchSize the size of the bulk to use
+    * @param successfulResults accumulator of successfull results
+    * @return Queried addresses
+    */
+  @tailrec
+  final def iterateOverRequestsWithBackPressure(requests: Stream[BulkAddressRequestData], miniBatchSize: Int,
+    successfulResults: Seq[BulkAddress]): Seq[BulkAddress] = {
+
+    val defaultBatchSize = conf.config.bulkRequestsPerBatch
+    val bulkSizeWarningThreshold = conf.config.bulkRequestsSizeWarningThreshold
+
+    if (miniBatchSize < defaultBatchSize * bulkSizeWarningThreshold)
+      logger.warn(s"#bulkQuery mini-bulk size it less than a ${defaultBatchSize * bulkSizeWarningThreshold}: size = $miniBatchSize , check if everything is fine with ES")
+    else logger.info(s"#bulkQuery sending a mini-batch of the size $miniBatchSize")
+
+    val miniBatch = requests.take(miniBatchSize)
+    val requestsAfterMiniBatch = requests.drop(miniBatchSize)
+    val result: BulkAddresses = Await.result(queryBulkAddresses(miniBatch, conf.config.bulkLimit), Duration.Inf)
+
+    val requestsLeft = result.failedRequests ++ requestsAfterMiniBatch
+
+    if (requestsLeft.isEmpty) successfulResults ++ result.successfulBulkAddresses
+    else if (miniBatchSize == 1 && result.failedRequests.nonEmpty) throw new Exception("Bulk query request: mini-bulk was scaled down to the size of 1 and it still fails, something's wrong with ES.")
```

Comment:
The exception from ES is already logged (as an `INFO`) and I would need to add a field to `BulkAddressRequestData` to store the "bubbled" exception message just to show it in this warning. It may be worth it, let me meditate on it.